aptos/nft_bridge/sources/contract_upgrade.move

```diff
@@ -0,0 +1,217 @@
+/// transaction, because the transaction that uploads bytecode cannot execute
+/// it.
+module nft_bridge::contract_upgrade {
+    use std::vector;
+    use aptos_framework::code;
+    use wormhole::deserialize;
+    use wormhole::cursor;
+    use wormhole::vaa;
+    use wormhole::state as core;
+    use wormhole::keccak256::keccak256;
+
+    use nft_bridge::vaa as nft_bridge_vaa;
+    use nft_bridge::state;
+
+    /// "NFTBridge" (left padded)
+    const NFT_BRIDGE: vector<u8> = x"00000000000000000000000000000000000000000000004e4654427269646765";
+
+    const E_UPGRADE_UNAUTHORIZED: u64 = 0;
+    const E_UNEXPECTED_HASH: u64 = 1;
+    const E_INVALID_MODULE: u64 = 2;
+    const E_INVALID_ACTION: u64 = 3;
+    const E_INVALID_TARGET: u64 = 4;
+    const E_NOT_MIGRATING: u64 = 5;
+
+    /// The `UpgradeAuthorized` type in the global storage represents the fact
+    /// there is an ongoing approved upgrade.
+    /// When the upgrade is finalised in `upgrade`, this object is deleted.
+    struct UpgradeAuthorized has key {
+        hash: vector<u8>
+    }
+
+    struct Hash has drop {
+        hash: vector<u8>
+    }
+
+    public fun get_hash(hash: &Hash): vector<u8> {
+        hash.hash
+    }
+
+    fun parse_payload(payload: vector<u8>): Hash {
+        let cur = cursor::init(payload);
+        let target_module = deserialize::deserialize_vector(&mut cur, 32);
+
+        assert!(target_module == NFT_BRIDGE, E_INVALID_MODULE);
+
+        let action = deserialize::deserialize_u8(&mut cur);
+        assert!(action == 0x02, E_INVALID_ACTION);
+
+        let chain = deserialize::deserialize_u16(&mut cur);
+        assert!(chain == core::get_chain_id(), E_INVALID_TARGET);
+
+        let hash = deserialize::deserialize_vector(&mut cur, 32);
+
+        cursor::destroy_empty(cur);
+
+        Hash { hash }
+    }
+
+// -----------------------------------------------------------------------------
+// Commit
+
+    public fun submit_vaa(vaa: vector<u8>): Hash acquires UpgradeAuthorized {
+        let vaa = vaa::parse_and_verify(vaa);
+        vaa::assert_governance(&vaa);
+        nft_bridge_vaa::replay_protect(&vaa);
+
+        let hash = parse_payload(vaa::destroy(vaa));
+        authorize_upgrade(&hash);
+        hash
+    }
+
+    public entry fun submit_vaa_entry(vaa: vector<u8>) acquires UpgradeAuthorized {
+        submit_vaa(vaa);
+    }
+
+    fun authorize_upgrade(hash: &Hash) acquires UpgradeAuthorized {
+        let nft_bridge = state::nft_bridge_signer();
+        if (exists<UpgradeAuthorized>(@nft_bridge)) {
+            // TODO(csongor): here we're dropping the upgrade hash, in case an
+            // upgrade fails for some reason. Should we emit a log or something?
+            let UpgradeAuthorized { hash: _ } = move_from<UpgradeAuthorized>(@nft_bridge);
+        };
+        move_to(&nft_bridge, UpgradeAuthorized { hash: hash.hash });
+    }
+
+    #[test_only]
+    public fun authorized_hash(): vector<u8> acquires UpgradeAuthorized {
```

Comment:
Is there a reason this is test_only?

Reply:
Just feels like a generally useful function for confirming when a upgrade vaa was successfully submitted...

Reply:
these functions can't be run off chain anyway, we can just look up the storage directly and query the auth object